prow/cmd/deck/main.go

```diff
@@ -1131,68 +1161,75 @@ func handleArtifactView(o options, sg *spyglass.Spyglass, cfg config.Getter) htt
+		handleRemoteLens(*lens, w, r, resource, request)
+	}
+}
+
+func handleRemoteLens(lens config.LensFileConfig, w http.ResponseWriter, r *http.Request, resource string, request spyglass.LensRequest) {
+	var requestType spyglassapi.RequestAction
+	switch resource {
+	case "iframe":
+		requestType = spyglassapi.RequestActionInitial
+	case "rerender":
+		requestType = spyglassapi.RequestActionRerender
+	case "callback":
+		requestType = spyglassapi.RequestActionCallBack
+	default:
+		http.NotFound(w, r)
+		return
+	}
+
+	var data string
+	if requestType != spyglassapi.RequestActionInitial {
+		dataBytes, err := ioutil.ReadAll(r.Body)
 		if err != nil {
-			http.Error(w, fmt.Sprintf("Failed to retrieve expected artifacts: %v", err), http.StatusInternalServerError)
+			http.Error(w, fmt.Sprintf("Failed to read body: %v", err), http.StatusInternalServerError)
 			return
 		}
+		data = string(dataBytes)
+	}
 
-		switch resource {
-		case "iframe":
-			t, err := template.ParseFiles(path.Join(o.templateFilesLocation, "spyglass-lens.html"))
-			if err != nil {
-				http.Error(w, fmt.Sprintf("Failed to load template: %v", err), http.StatusInternalServerError)
-				return
-			}
-
-			w.Header().Set("Content-Type", "text/html; encoding=utf-8")
-			t.Execute(w, struct {
-				Title   string
-				BaseURL string
-				Head    template.HTML
-				Body    template.HTML
-			}{
-				lensConfig.Title,
-				"/spyglass/static/" + lensName + "/",
-				template.HTML(lens.Header(artifacts, lensResourcesDir, cfg().Deck.Spyglass.Lenses[request.Index].Lens.Config)),
-				template.HTML(lens.Body(artifacts, lensResourcesDir, "", cfg().Deck.Spyglass.Lenses[request.Index].Lens.Config)),
-			})
-		case "rerender":
-			data, err := ioutil.ReadAll(r.Body)
-			if err != nil {
-				http.Error(w, fmt.Sprintf("Failed to read body: %v", err), http.StatusInternalServerError)
-				return
-			}
-			w.Header().Set("Content-Type", "text/html; encoding=utf-8")
-			w.Write([]byte(lens.Body(artifacts, lensResourcesDir, string(data), cfg().Deck.Spyglass.Lenses[request.Index].Lens.Config)))
-		case "callback":
-			data, err := ioutil.ReadAll(r.Body)
-			if err != nil {
-				http.Error(w, fmt.Sprintf("Failed to read body: %v", err), http.StatusInternalServerError)
-				return
-			}
-			w.Write([]byte(lens.Callback(artifacts, lensResourcesDir, string(data), cfg().Deck.Spyglass.Lenses[request.Index].Lens.Config)))
-		default:
-			http.NotFound(w, r)
-		}
+	lensRequest := spyglassapi.LensRequest{
+		Action:         requestType,
+		Data:           data,
+		Config:         lens.Lens.Config,
+		ResourceRoot:   "/spyglass/static/" + lens.Lens.Name + "/",
+		Artifacts:      request.Artifacts,
+		ArtifactSource: request.Source,
+		LensIndex:      request.Index,
+	}
+	serializedRequest, err := json.Marshal(lensRequest)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to marshal request to lens backend: %v", err), http.StatusInternalServerError)
+		return
 	}
+
+	(&httputil.ReverseProxy{
```

Comment:
This is weird enough to merit some eyeballs -- so this shares a lifetime with the goroutine, right? Does it leak memory?

Reply:
Yes it shares a lifetime with the goroutine, I don't think this leaks memory

Reply:
(I've burned my fingers already in the past by assuming this can be implemented by doing a simple io.Copy on the response, but turns out, HTTP is more complicated than that so I've refrained from trying to be smart)